src/libraries/System.Linq/src/System/Linq/Take.cs

```diff
@@ -20,6 +20,169 @@ public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> sourc
                 TakeIterator<TSource>(source, count);
         }
 
+        /// <summary>Returns a specified range of contiguous elements from a sequence.</summary>
+        /// <param name="source">The sequence to return elements from.</param>
+        /// <param name="range">The range of elements to return, which has start and end indexes either from the start or the end.</param>
+        /// <typeparam name="TSource">The type of the elements of <paramref name="source" />.</typeparam>
+        /// <exception cref="ArgumentNullException">
+        ///   <paramref name="source" /> is <see langword="null" />.
+        /// </exception>
+        /// <returns>An <see cref="IEnumerable{T}" /> that contains the specified <paramref name="range" /> of elements from the <paramref name="source" /> sequence.</returns>
+        public static IEnumerable<TSource> Take<TSource>(this IEnumerable<TSource> source, Range range)
+        {
+            if (source == null)
+            {
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
+            }
+
+            Index start = range.Start;
+            Index end = range.End;
+            bool isStartIndexFromEnd = start.IsFromEnd;
+            bool isEndIndexFromEnd = end.IsFromEnd;
+            int startIndex = start.Value;
+            int endIndex = end.Value;
+            Debug.Assert(startIndex >= 0);
+            Debug.Assert(endIndex >= 0);
+
+            if (!isStartIndexFromEnd && !isEndIndexFromEnd)
+            {
+                return startIndex >= endIndex
+                    ? Empty<TSource>()
+                    : source switch
+                    {
+                        IPartition<TSource> partition => partition.Skip(startIndex).Take(endIndex - startIndex),
+                        IList<TSource> list => new ListPartition<TSource>(list, startIndex, endIndex - 1),
+                        _ => new EnumerablePartition<TSource>(source, startIndex, endIndex - 1)
+                    };
+            }
```

Comment:
This is going to break once https://github.com/dotnet/runtime/pull/47918 goes in.

Can we instead just make it:
```C#
if (!isStartIndexFromEnd && !isEndIndexFromEnd)
{
    return startIndex >= endIndex ?
        Empty<TSource>() :
        source.Skip(startIndex).Take(endIndex - startIndex);
}
```
?

Reply:
Updated.